File: android/app/src/main/java/com/alexivanov/snapsell/domain/ListingEnums.kt
package com.alexivanov.snapsell.domain

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/** Matches contracts/listing.schema.json "status". */
@Serializable
enum class ListingStatus(val wire: String, val label: String) {
    @SerialName("draft") DRAFT("draft", "Draft"),
    @SerialName("listed") LISTED("listed", "Listed"),
    @SerialName("skipped") SKIPPED("skipped", "Skipped"),
    @SerialName("sold") SOLD("sold", "Sold");

    companion object {
        fun fromWire(value: String): ListingStatus =
            entries.firstOrNull { it.wire == value }
                ?: throw IllegalArgumentException("Unknown listing status: $value")
    }
}

/** Matches contracts/listing.schema.json "kind". */
@Serializable
enum class ListingKind(val wire: String, val label: String) {
    @SerialName("single") SINGLE("single", "Single"),
    @SerialName("bundle") BUNDLE("bundle", "Bundle");

    companion object {
        fun fromWire(value: String): ListingKind =
            entries.firstOrNull { it.wire == value }
                ?: throw IllegalArgumentException("Unknown listing kind: $value")
    }
}

File: ios/SnapSell/Domain/ListingEnums.swift
import Foundation

/// Matches contracts/listing.schema.json "status".
enum ListingStatus: String, Codable, CaseIterable {
    case draft, listed, skipped, sold

    var wire: String { rawValue }
    var label: String { rawValue.prefix(1).uppercased() + rawValue.dropFirst() }

    static func fromWire(_ value: String) throws -> ListingStatus {
        guard let s = ListingStatus(rawValue: value) else { throw AppError("Unknown listing status: \(value)") }
        return s
    }
}

/// Matches contracts/listing.schema.json "kind".
enum ListingKind: String, Codable, CaseIterable {
    case single, bundle

    var wire: String { rawValue }
    var label: String { rawValue.prefix(1).uppercased() + rawValue.dropFirst() }

    static func fromWire(_ value: String) throws -> ListingKind {
        guard let k = ListingKind(rawValue: value) else { throw AppError("Unknown listing kind: \(value)") }
        return k
    }
}
